### PR TITLE
Bazel: minimal test build transitions

### PR DESCRIPTION
**Pull request by prestonvanloon**
**What type of PR is this?**

Other

**What does this PR do? Why is it needed?**

Adds build transitions to support `bazel build //...` with mix build configs.

**Which issues(s) does this PR fix?**

**Other notes for review**

Replaces the `--define=ssz=minimal` with a more standard practice for config settings. Going forward, bazel users should use `--//proto:network=<value>` to specify the configuration setting for build time templating in protobuf definitions. One of the benefits of this change is that we can now hardcode configurations within starlark rules. Our custom `go_test` definition is an extension of the upstream `go_test` with an incoming edge transition that sets the config and gotags to the appropriate network value. In other words, if a go_test has `eth_network = "minimal"`, bazel will insure that all incoming dependencies will be compiled with the "minimal" go tag and "minimal" ssz config protobufs.

Examples: 
- `//endtoend:go_default_test` no longer requires `--define=ssz=minimal` or the new flag `--proto:network=minimal`, instead, the incoming edge transition handles this automatically and users simply need to run `bazel test //endtoend:go_default_test`. 
- `bazel test //spectest/...` now runs all mainnet and minimal spec tests in a single command.

### CODE REVIEW

**Comment by prestonvanloon** [comment]
Note for reviewers, remove display of BUILD files. 
Screenshot from 2021-06-29 17-52-15

File: tools/go/def.bzl [hunk 1]
```diff
@@ -1,6 +1,51 @@
...
+        "eth_network": attr.string(values = ["mainnet", "minimal"]),
+        "_whitelist_function_transition": attr.label(
+            default = "@bazel_tools//tools/whitelists/function_transition_whitelist",
+        ),
+    })
+    kwargs["attrs"] = attrs
+    kwargs["cfg"] = go_test_transition
+    return rule(**kwargs)
+
+go_test = _go_test_transition_rule(**go_test_kwargs)
```
Comment (prestonvanloon) [inline tools/go/def.bzl:47]: This is the meat of the PR logic. 

On line 47, we define a new `go_test` which uses the same kwargs as go_test with some additions.

File: third_party/io_bazel_rules_go_test.patch [hunk 2]
```diff
@@ -0,0 +1,24 @@
+diff --git a/go/private/rules/test.bzl b/go/private/rules/test.bzl
+index a7936ca9..cbfc996a 100644
+--- a/go/private/rules/test.bzl
++++ b/go/private/rules/test.bzl
+@@ -176,7 +176,7 @@ def _go_test_impl(ctx):
+         ),
+     ]
+ 
+-_go_test_kwargs = {
++go_test_kwargs = {
```
Comment (prestonvanloon) [inline third_party/io_bazel_rules_go_test.patch:10]: This just makes a private variable public.

File: WORKSPACE [hunk 3]
```diff
@@ -88,6 +88,8 @@ http_archive(
         # Required until https://github.com/bazelbuild/rules_go/pull/2450 merges otherwise nilness
         # nogo check fails for certain third_party dependencies.
         "//third_party:io_bazel_rules_go.patch",
+        # Expose internals of go_test for custom build transitions.
+        "//third_party:io_bazel_rules_go_test.patch",
```
Comment (prestonvanloon) [inline WORKSPACE:92]: Patch to expose private variables in rules_go

File: endtoend/BUILD.bazel [hunk 4]
```diff
@@ -20,6 +20,7 @@ go_test(
         "//tools/bootnode",
         "@com_github_ethereum_go_ethereum//cmd/geth",
     ],
+    eth_network = "minimal",
```
Comment (prestonvanloon) [inline endtoend/BUILD.bazel:23]: Example of defining a minimal network build configuration

**Review by rauljordan — APPROVED** [review]

**Review by rauljordan — APPROVED** [review]
Reviewed and I understand how to use the rule. This will be super helpful for everyone moving forward

**Review by rauljordan — APPROVED** [review]
Reviewed carefully and I understand how this new rule works. Thanks for doing this! It will be awesome to make e2e this easy to run

**Pull request merged**